webrender/src/font_renderer.rs

```diff
@@ -0,0 +1,367 @@
+        }
+    }
+
+    /// Get access to the font context usable outside of the thread pool.
+    pub fn lock_shared_context(&self) -> MutexGuard<Context> {
+        self.shared_context.lock().unwrap()
+    }
+
+    // number of contexts associated to workers
+    pub fn num_worker_contexts(&self) -> usize {
+        self.worker_contexts.len()
+    }
+
+    fn current_worker_id(&self) -> Option<usize> {
+        self.workers.current_thread_index()
+    }
+}
+
+pub struct FontRenderer {
+    workers: Arc<ThreadPool>,
+    font_contexts: Arc<FontContexts>,
+
+    // Receives the rendered glyphs.
+    glyph_rx: Receiver<GlyphRasterJob>,
+
+    // Maintain a set of glyphs that have been requested this
+    // frame. This ensures the glyph thread won't rasterize
+    // the same glyph more than once in a frame. This is required
+    // because the glyph cache hash table is not updated
+    // until the end of the frame when we wait for glyph requests
+    // to be resolved.
+    pending_glyphs: HashSet<GlyphRequest>,
+
+    // We defer removing fonts to the end of the frame so that:
+    // - this work is done outside of the critical path,
+    // - we don't have to worry about the ordering of events if a font is used on
+    //   a frame where it is used (although it seems unlikely).
+    fonts_to_remove: Vec<FontKey>,
+}
+
+impl FontRenderer {
+    pub fn new(workers: Arc<ThreadPool>) -> Self {
+        let (glyph_tx, glyph_rx) = channel();
+
+        let num_workers = workers.current_num_threads();
+        let mut contexts = Vec::with_capacity(num_workers);
+
+        for _ in 0..num_workers {
+            contexts.push(
+                Mutex::new(
+                    Context {
+                        context: FontContext::new(),
+                        tx: glyph_tx.clone(),
+                    }
+                )
+            );
+        }
+
+        FontRenderer {
+            font_contexts: Arc::new(
+                FontContexts {
+                    worker_contexts: contexts,
+                    shared_context: Mutex::new(
+                        Context {
+                            context: FontContext::new(),
+                            tx: glyph_tx.clone(),
+                        }
+                    ),
+                    workers: Arc::clone(&workers),
+                }
+            ),
+            glyph_rx: glyph_rx,
+            pending_glyphs: HashSet::new(),
+            workers: workers,
+            fonts_to_remove: Vec::new(),
+        }
+    }
+
+    pub fn add_font(&mut self, font_key: FontKey, template: FontTemplate) {
+        let font_contexts = Arc::clone(&self.font_contexts);
+        // It's important to synchronously add the font for the shared context because
+        // we use it to check that fonts have been properly added when requesting glyphs.
+        font_contexts.lock_shared_context().add_font(font_key, &template);
+        self.workers.spawn_async(move || {
```

Comment:
Note: I am thinking of just doing this from the render backend thread instead of spawning an async job, in order to keep it simple. FontContext::add_font should be cheap (and if it is not we can easily make it so), and that way we have the guarantee that fonts are added for the subsequent glyph requests without relying on any ordering in the thread pool. otherwise we need to account for the possibility that a the thread pool process tasks out of order and run rasterize_glyph before adding the font.